sheepdog/transactions/upload/__init__.py

```diff
@@ -32,8 +36,27 @@ def single_transaction_worker(transaction, *doc_args):
             transaction.flush()
             transaction.post_validate()
             transaction.commit()
-        except HandledIntegrityError:
-            pass
+        except IntegrityError:
+            transaction.session.rollback()
+            for entity in transaction.valid_entities:
+                schema = gdcdictionary.schema[entity.node.label]
+                node = entity.node
+                for keys in schema['uniqueKeys']:
+                    props = {}
+                    if keys == ['id']:
+                        continue
+                    for key in keys:
+                        prop = schema['properties'][key].get('systemAlias')
+                        if prop:
+                            props[prop] = node[prop]
+                        else:
+                            props[key] = node[key]
+                    if transaction.db_driver.nodes(type(node)).props(props).count() > 0:
+                            entity.record_error(
+                                '{} with {} already exists in the DB'
+                                .format(node.label, props), keys=props.keys()
+                            )
```

Comment:
It would be nice to make this handler a method of `transaction`. The test wouldn't need to copy the code too.